src/python/pants/rules/core/test.py

```diff
@@ -141,9 +158,23 @@ class AddressAndDebugRequest:
 
   results = await MultiGet(Get[AddressAndTestResult](Address, addr.to_address()) for addr in addresses)
   did_any_fail = False
-  filtered_results = [(x.address, x.test_result) for x in results if x.test_result is not None]
+  filtered_address_and_test_results = tuple(x for x in results if x.test_result is not None)
+  if options.values.run_coverage:
+    # TODO(#8915) Make a generic interface for collecting coverage reports. This class shouldn't have to know about
+    # PytestCoverageReport and JunitCoverageReport etc. It should only know about CoverageReport.
+    pytest_coverage_report = await Get[PytestCoverageReport](
+      AddressAndTestResults,
+      AddressAndTestResults(filtered_address_and_test_results)
+    )
+    workspace.materialize_directory(DirectoryToMaterialize(
+      pytest_coverage_report.report_directory_digest,
+      path_prefix=str(pytest_coverage_report.directory_to_materialize_to)
+    ))
+    console.print_stdout(f"Wrote coverage report to `{pytest_coverage_report.directory_to_materialize_to}`")
 
+  filtered_results = tuple((x.address, x.test_result) for x in filtered_address_and_test_results)
   for address, test_result in filtered_results:
+    assert test_result is not None
```

Comment:
Instead, add `if test_result is not None` to line 175 in the generator expression.